docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+containers (they are run serially, rather than in parallel), for backwards
+compatibility with existing clients they must be identified as distinct
+fields in the API:
+
+    pod:
+      spec:
+        containers: ...
+        initContainers:
+        - name: init-container1
+          image: ...
+          ...
+        - name: init-container2
+        ...
+      status:
+        containerStatuses: ...
+        initContainerStatuses:
+        - name: init-container1
+          ...
+        - name: init-container2
+          ...
+
+This separation also serves to make the order of container initialization
+clear - init containers are executed in the order that they appear, then all
+app containers are started at once.
+
+The name of each app and init container in a pod must be unique - it is a
+validation error for any container to share a name.
+
+
+### Resources
+
+Given the ordering and execution for init containers, the following rules
+for resource usage apply:
+
+* The highest of any particular resource request or limit defined on all init
+  containers is the **effective init request/limit**
+* The pod's **effective request/limit** for a resource is the higher of:
+  * sum of all app containers request/limit for a resource
+  * effective init request/limit for a resource
+* The highest QoS tier of init containers is the **effective init QoS tier**
+
+So the following pod:
+
+    pod:
+      spec:
+        initContainers:
+        - limits:
+				    cpu: 100m
+				    memory: 1GiB
+        - limits:
+				    cpu: 50m
+				    memory: 2GiB
+			  containers:
+        - limits:
+				    cpu: 10m
+				    memory: 1100MiB
+        - limits:
+				    cpu: 10m
+				    memory: 1100MiB
+
+has an effective pod limit of `cpu: 100m`, `memory: 2200MiB` (highest init
+container cpu is larger than sum of all app containers, sum of container
+memory is larger than the max of all init containers). The scheduler, node,
+and quota must respect the effective pod request/limit.
+
+
+### Kubelet and container runtime details
+
+Container runtimes should treat the set of init and app containers as one
+large pool. An individual init container execution should be identical to
+an app container.
+
+All app container operations are permitted on init containers. The
+logs for an init container should be available for the duration of the pod
+lifetime or until the pod is restarted.
+
+During initialization, app container status should be shown with the reason
+PodInitializing if any init containers are present. Each init container
+should show appropriate container status, and all init containers that are
+waiting for earlier init containers to finish should have the `reason`
+PendingInitialization.
+
+The container runtime should aggressively prune failed init containers.
+The container runtime should record whether all init containers have
+succeeded internally, and only invoke new init containers if a pod
```

Comment:
Yes, to make it clear (there has been no disagreement so far that they can
be executed multiple times).

On Tue, Apr 12, 2016 at 5:58 PM, Vish Kannan <notifications@github.com>
wrote:

> In docs/proposals/container-init.md
> <https://github.com/kubernetes/kubernetes/pull/23666#discussion_r59461605>
> :
>
> > +large pool. An individual init container execution should be identical to
> > +an app container.
> > +
> > +All app container operations are permitted on init containers. The
> > +logs for an init container should be available for the duration of the pod
> > +lifetime or until the pod is restarted.
> > +
> > +During initialization, app container status should be shown with the reason
> > +PodInitializing if any init containers are present. Each init container
> > +should show appropriate container status, and all init containers that are
> > +waiting for earlier init containers to finish should have the `reason`
> > +PendingInitialization.
> > +
> > +The container runtime should aggressively prune failed init containers.
> > +The container runtime should record whether all init containers have
> > +succeeded internally, and only invoke new init containers if a pod
>
> @smarterclayton <https://github.com/smarterclayton> Should we mention the
> fact that init containers have to be idempotent explicitly?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23666/files/7efdc3d36197f1ca452339d49abbc87163fb0e75#r59461605>
>